cni/pkg/taint/taint.go

```diff
@@ -0,0 +1,387 @@
+package taint
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	client "k8s.io/client-go/kubernetes"
+	"k8s.io/kubectl/pkg/util/podutils"
+
+	"istio.io/pkg/log"
+)
+
+const (
+	TaintName     = "NodeReadiness"
+	ToleranceName = "NodeReadiness"
+)
+
+type ConfigSettings struct {
+	name           string
+	namespace      string
+	LabelSelectors string `json:"label_selectors"`
+}
+
+func (config ConfigSettings) ToString() string {
+	return fmt.Sprintf("name is %s, namespace is %s,  selector is %s", config.name, config.namespace, config.LabelSelectors)
+}
+
+type Options struct {
+	ConfigName      string
+	ConfigNameSpace string
+}
+type TaintSetter struct {
+	configs []ConfigSettings // contains all configmaps information
+	Client  client.Interface
+}
+
+func (ts TaintSetter) GetAllConfigs() []ConfigSettings {
+	return ts.configs
+}
+func NewTaintSetter(clientset client.Interface, options *Options) (ts TaintSetter, err error) {
+	configmap, err := clientset.CoreV1().ConfigMaps(options.ConfigNameSpace).Get(context.TODO(), options.ConfigName, metav1.GetOptions{})
+	if err != nil {
+		return ts, err
+	}
+	ts = TaintSetter{
+		configs: []ConfigSettings{},
+		Client:  clientset,
+	}
+	ts.LoadConfig(*configmap)
+	return
+}
+
+// load corresponding configmap's critical labels and their namespace
+func (ts *TaintSetter) LoadConfig(config v1.ConfigMap) {
+	log.Debugf("Loading configmap %s in %s", config.Name, config.Namespace)
+	ts.configs = make([]ConfigSettings, 0) //clear previous one
+	for key, value := range config.Data {
+		log.Debugf("Loading %s ", key)
+		temp := strings.Split(value, "\n")
+		var tempset ConfigSettings
+		for _, str := range temp {
+			if strings.HasPrefix(strings.TrimSpace(str), "name=") {
+				pair := strings.SplitN(strings.TrimSpace(str), "=", 2)
+				tempset.name = pair[1]
+			} else if strings.HasPrefix(strings.TrimSpace(str), "namespace=") {
+				pair := strings.SplitN(strings.TrimSpace(str), "=", 2)
+				tempset.namespace = pair[1]
+			} else if strings.HasPrefix(strings.TrimSpace(str), "selector=") {
+				pair := strings.SplitN(strings.TrimSpace(str), "=", 2)
+				tempset.LabelSelectors = pair[1]
+			}
+		}
+		ts.configs = append(ts.configs, tempset)
+		fmt.Println(ts.configs)
+	}
+}
+func (ts TaintSetter) validTolerance(pod v1.Pod) bool {
+	for _, toleration := range pod.Spec.Tolerations {
+		if (toleration.Key == ToleranceName || toleration.Key == "") &&
+			toleration.Operator == v1.TolerationOpExists &&
+			toleration.Effect == v1.TaintEffectNoSchedule {
+			return true
+		}
+	}
+	return false
+}
+
+// find all pods with selectors and namespaces defined in config map
+func (ts TaintSetter) ListCandidatePods() (list v1.PodList, err error) {
```

Comment:
Add parameters `nodeName`, `podNamespace`, and `labelSelector` here. Then you can just query for e.g. `ListCandidatePods("node01", "istio-system", "app:istio-cni")` or similar and get a list of pods in the istio-system namespace running on node01 that match a given label.